File: BFSTraversal.java
    /*  Explanation
    # Leetcode problem link : https://leetcode.com/problems/binary-tree-level-order-traversal/
    Time Complexity for operators : o(n) .. number of nodes in tree.
    Extra Space Complexity for operators : o(n) ... used queue
    Did this code successfully run on Leetcode : Yes
    Any problem you faced while coding this : No
# Your code here along with comments explaining your approach
        # Basic approach : 

        # Optimized approach same like basic approach: Recursion or Iterative.
                              
            Iterative approach   
                    A. Add first element into the queue as root.
                    B. Traverse thru the queue until it becomes empty.
                    C. In while loop, check the size of queue first that gives the size of level.
                    D. Traverse thru the for loop and add all the elements till that size.
                    E. check if that node has left or right nodes if it has then add that into the queue.
                    F. after for loop, add that list into the final outout.
                    G. Return final output.

       */


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

   class BFSTraversal {

    public List<List<Integer>> levelOrder(TreeNode root) {
        
        if(root==null)
            return new ArrayList<>();
        
        List<List<Integer>> result = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        
        queue.add(root);
        while(!queue.isEmpty()){
            int size = queue.size();
            List<Integer> list = new ArrayList<>();

            for(int i=0;i<size;i++){
                TreeNode tr = queue.poll();
                
                list.add(tr.val);
                
                if(tr.left!=null) queue.add(tr.left);
                if(tr.right!=null) queue.add(tr.right);
            }
            
            result.add(list);
        }
        
        return result;
    }
}

File: bfstraversal.py
# leetcode: accepted
# no doubts

# you maintain level. If the level is eqUAL TO THE ;EN OF THE LIST, YOU APPEND A NEW LIST AND THE VALUE OF THE ROOT TO THE LIST[0].
# if the level is greater tham the len, the index'th list is already presnent, just append the values


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        result = []  # creating the resultant list
        level = 0  # initalising the lvel has 0
        self.helper(root, level, result)
        return result

    def helper(self, root, level, result):
        if root == None:  # if the root is null return
            return
        if level == len(result):  # if the current level equals the lenght of the result, yu add a new list
            result.append(list())  # adding a list
        result[level].append(root.val)  # appending the value to the result[level]

        self.helper(root.left, level + 1, result)  # recursively caling the left child, and increasing level
        self.helper(root.right, level + 1, result)  # recursively caling the right child and and increasing level